[excel, CommonAPI] (Code patterns) Adding notes about range limits and promiseless callback methods

AlexJerabek opened the pull request:
This adds a section centralizing the limitations with Excel range sizes. It also discusses the Common/Outlook programming model and its lack of promises.

o365devx commented:
Docs Build status updates of commit _[049d37e](https://github.com/OfficeDev/office-js-docs-pr/commits/049d37ea0316793ba90d3dc5f3314aeba50fe04b)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/common-coding-issues.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-RangeLimits/docs/develop/common-coding-issues.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C10%5C30%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201910302111106338-1406%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=ZmwDhB%2BAf3BNXKUkCGK2MUKSAsF0afJLC0eRWc1NpNQ%3D&st=2019-10-30T21%3A07%3A17Z&se=2019-11-30T21%3A12%3A17Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

kbrandl commented on docs/develop/common-coding-issues.md:
```diff
@@ -35,6 +41,17 @@ You can identify a property that must have its subproperties set with a JSON str
 - Read-only property: Subproperties can be set through navigation.
 - Writable property: Subproperties must be set with a JSON struct (and cannot be set through navigation).
 
+## Excel Range limits
```
I'm wondering if (for discoverability), it'd be better to add this information to the [Ranges (advanced)](https://docs.microsoft.com/en-us/office/dev/add-ins/excel/excel-add-ins-ranges-advanced) article, instead of here in this coding issues article?  Or, perhaps define the contents of the section in an include file, and include it both here in the coding issues article and also in the Excel Ranges (advanced) article? Thoughts?

AlexJerabek replied:
There's a section in the Excel Core Concepts doc that mentions the 5MB limit. I'll replace that with a link to this section. I'll also add a link to the new coding issues page from the **See Also** list.

kbrandl commented on docs/develop/common-coding-issues.md:
```diff
@@ -1,14 +1,20 @@
 
+## Common and Outlook APIs are not promise-based
+
+The [Common APIs](/javascript/api/office) (those that are not tied to a particular host) and [Outlook APIs](/javascript/api/outlook) use a callback-based programming model. Interacting with the underlying Office document requires an asynchronous read or write call that includes a callback to be ran when the operation completes. For an example of this pattern, see [AppointmentCompose.getAttachment ...
```
```suggestion
The [Common APIs](/javascript/api/office) (those that are not tied to a particular Office host) and [Outlook APIs](/javascript/api/outlook) use a callback-based programming model. Interacting with the underlying Office document requires an asynchronous read or write call that specifies a callback to be ran when the operation completes. For an example of this pattern, see [AppointmentCompose.getAttachmentAsync](/javascript/api/office/office.file?view=common-js#getsliceasync-sliceindex--callback-).
```

kbrandl replied:
The URL target (bookmark `#getsliceasync-sliceindex--callback-`) specified at the end of this paragraph for the link `AppointmentCompose.getAttachmentAsync` doesn't point to the right place in the reference docs.

kbrandl commented on docs/develop/common-coding-issues.md:
```diff
@@ -1,14 +1,20 @@
 
 This article highlights aspects of the Office JavaScript API that may result in unexpected behavior or require specific coding patterns to achieve the desired outcome. If you encounter an issue that belongs in this list, please let us know by using the feedback form at the bottom of the article.
 
+## Common and Outlook APIs are not promise-based
```
```suggestion
## Common APIs and Outlook APIs are not promise-based
```

kbrandl commented on docs/develop/common-coding-issues.md:
```diff
@@ -1,14 +1,20 @@
+
+The [Common APIs](/javascript/api/office) (those that are not tied to a particular host) and [Outlook APIs](/javascript/api/outlook) use a callback-based programming model. Interacting with the underlying Office document requires an asynchronous read or write call that includes a callback to be ran when the operation completes. For an example of this pattern, see [AppointmentCompose.getAttachment ...
+
+These Common and Outlook methods do not return [Promises](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise). As such, you cannot use [await](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/await) to pause the execution until the async operation completes. If you need `await` behavior, you can wrap the method call in an explicitly created Pro ...
```
```suggestion
These Common API methods and Outlook API methods do not return [Promises](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise). Therefore, you cannot use [await](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/await) to pause the execution until the asynchronous operation completes. If you need `await` behavior, you can wrap the method call in an explicitly created Promise.
```

kbrandl commented on docs/develop/common-coding-issues.md:
```diff
@@ -1,14 +1,20 @@
+
+The [Common APIs](/javascript/api/office) (those that are not tied to a particular host) and [Outlook APIs](/javascript/api/outlook) use a callback-based programming model. Interacting with the underlying Office document requires an asynchronous read or write call that includes a callback to be ran when the operation completes. For an example of this pattern, see [AppointmentCompose.getAttachment ...
+
+These Common and Outlook methods do not return [Promises](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise). As such, you cannot use [await](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/await) to pause the execution until the async operation completes. If you need `await` behavior, you can wrap the method call in an explicitly created Pro ...
```
Would it be helpful to include a code sample here that demonstrates this approach?

AlexJerabek replied:
Probably. I'll craft and test one.

kbrandl commented on docs/develop/common-coding-issues.md:
```diff
@@ -35,6 +41,17 @@ You can identify a property that must have its subproperties set with a JSON str
 
+## Excel Range limits
+
+Excel developers need to be aware of two data size limitations:
```
```suggestion
If you're building an Excel add-in that uses ranges, be aware of the following size limitations:
```

kbrandl commented on docs/develop/common-coding-issues.md:
```diff
@@ -35,6 +41,17 @@ You can identify a property that must have its subproperties set with a JSON str
+- Excel on the web has a payload size limit for requests and responses of 5MB. `RichAPI.Error` will be thrown if that limit is exceeded.
+- A range is limited to five million cells for set operations.
+
+If you expect user input to exceed those amounts, be sure to check the data and split the ranges into multiple objects. You'll also need separate `context.sync()` calls to avoid overwhelming the Excel cloud service.
```
```suggestion
If you expect user input to exceed these limits, be sure to check the data and split the ranges into multiple objects. You'll also need to submit multiple `context.sync()` calls to optimize performance.
```

AlexJerabek replied:
I'm taking the first two changes. To the third point, the issue isn't performance. They'll need to use multiple syncs to avoid the size limit error. I'll try to highlight that point better.

o365devx commented:
Docs Build status updates of commit _[e5dd6a2](https://github.com/OfficeDev/office-js-docs-pr/commits/e5dd6a2407240be00255b1caed687d488b488bb2)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/common-coding-issues.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-RangeLimits/docs/develop/common-coding-issues.md) | :white_check_mark:Succeeded | |
[docs/excel/excel-add-ins-core-concepts.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-RangeLimits/docs/excel/excel-add-ins-core-concepts.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C10%5C30%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201910302200051338-1406%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=eyfX1fSMJCBiNq8muENGwswYFt0tplOLffgz%2BZiLqLU%3D&st=2019-10-30T21%3A56%3A13Z&se=2019-11-30T22%3A01%3A13Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

o365devx commented:
Docs Build status updates of commit _[dba6d7c](https://github.com/OfficeDev/office-js-docs-pr/commits/dba6d7cfb5cc56ac5a6534b2d6365a23f5916211)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/common-coding-issues.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-RangeLimits/docs/develop/common-coding-issues.md) | :white_check_mark:Succeeded | |
[docs/excel/excel-add-ins-core-concepts.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-RangeLimits/docs/excel/excel-add-ins-core-concepts.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C10%5C31%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201910311959112654-1406%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=vLBBwxximEdyWOGfn%2FQd2K4bdCJHxQP7RW7ngRR0oms%3D&st=2019-10-31T19%3A55%3A17Z&se=2019-12-01T20%3A00%3A17Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

AlexJerabek commented:
Thanks for the review @kbrandl. I believe I've addressed the all the feedback. Also, thanks to @deeagar for the sample code (https://github.com/OfficeDev/office-js-docs-pr/issues/1404). Having external contributions really helps make these docs relevant to everyone.
